### PR TITLE
amp-graphiq: resize message fix and frozen support

### PR DESCRIPTION
**Pull request by iampueroo**
Hey! @src-code I added three commit listed here:

- **[Commit 1](https://github.com/src-code/amphtml/commit/eae4d862cfa283b98154bb79d606cd45b8a1d9af)**
   - There was a test failing when I ran it on init. It expected null but was undefined. Since the tested value wasn't from `amp-graphiq` I changed the test. 

- **[Commit 2](https://github.com/src-code/amphtml/commit/e5f53b4fad49f3423cdb178d3e416b1ceb4dc76c)**
    
   - Added a `&data-amp-version=true` to the request URL to signify that this is an `amp-graphiq` request. This mimics what we do for `script` style embeds which allows us to dynamically resize the element (the change will be deployed tomorrow).
  - Changed the message to the expect `{sentinel: "amp", type: "embed-size"}` format. I also used the `data-amp-version` flag to allow the widget to dynamically resize and message out a height "greater than what the widget was configured for". It resizes nicely, however I do see the `Cannot resize element and overflow is not available` warning.

- **[Commit 3](https://github.com/src-code/amphtml/commit/966ec6f0ab10ad0c17669d1bbcfdb86016602f39)** 
  - Adds support for "frozen" widgets. I'm hesistant about this addition, tbh, since it's far more internal to us (Graphiq). "Frozen" widgets are not live updating - we "freeze" the content of a widget at a point in time and point the iFrame `src` to it (`sw.graphiq.com`). This allows publishers to embed a widget about an event and make sure that the widget reflects what happened then, even months and years later. (Example: a stock crashes on a certain day, the publisher would "freeze" the stock chart so it doesn't update past that day/week). The issue is that you can't just freeze a widget by adding the `data-frozen` attribute to `<amp-graphiq>` - Graphiq has to actually freeze it. (And we do since we trigger a job when giving out the embed codes). Thus it's completely fine if Graphiq makes the frozen `<amp-graphiq>`s, but not if it's supposed to be a public attribute that anyone can use. Let me know what you think, I can revert the commit if it doesn't fit what these extensions should be.
    
Phew, that's it. Let me know your thoughts/what should change.

### CODE REVIEW

File: extensions/amp-graphiq/0.1/amp-graphiq.js [hunk 1]
```diff
@@ -70,6 +73,8 @@ class AmpGraphiq extends AMP.BaseElement {
 
   /** @override */
   buildCallback() {
+    const isFrozen = this.element.getAttribute('data-frozen');
```
Comment (jridgewell) [inline extensions/amp-graphiq/0.1/amp-graphiq.js]: This means `<amp-graphiq data-frozen>` isn't acceptable. Maybe use `#hasAttribute` instead?

**Comment by src-code** [comment]
This is fantastic, @iampueroo!  I'll add a few comments directly to the code.

As far as the frozen attribute goes, it does feel a little odd.  If I understand correctly, to be properly frozen the widget must be both frozen in the widget's configuration AND must be loaded from `sw.graphiq.com`?  So for a given widget id, could you load either the frozen or unfrozen version of it at any given time based on what domain you load it from, assuming that it's been frozen in configuration?  (So there's no 1:1 mapping between widget id and frozen state?)  If so, it sounds like we can't get around the need for that attribute.  (And really, even if a given widget id could only have one state, we still wouldn't know what domain to use on the client side without querying Graphiq for that state first, which isn't ideal.)  So it seems to be that the attribute would have to remain unless there was some way to for the client to discern the state automatically...

File: extensions/amp-graphiq/0.1/test/test-amp-graphiq.js [hunk 2]
```diff
@@ -82,31 +85,35 @@ describe('amp-graphiq', () => {
...
+    return getGraphiq(widgetId).then(graphiq => {
       const placeholder = graphiq.querySelector('[placeholder]');
-      testIframe(graphiq.querySelector('iframe'));
+      testIframe(graphiq.querySelector('iframe'), widgetId);
       const obj = graphiq.implementation_;
       obj.unlayoutCallback();
       expect(graphiq.querySelector('iframe')).to.be.null;
       expect(obj.iframe_).to.be.null;
-      expect(obj.iframePromise_).to.be.null;
+      expect(obj.iframePromise_).to.not.exist;
```
Comment (src-code) [inline extensions/amp-graphiq/0.1/test/test-amp-graphiq.js]: I think this is actually left over from a previous iteration where I stored an iframe promise in the component.  This test should be removed.

File: extensions/amp-graphiq/0.1/amp-graphiq.js [hunk 3]
```diff
@@ -136,7 +142,7 @@ class AmpGraphiq extends AMP.BaseElement {
     if (data === undefined) {
       return; // We only process valid JSON.
     }
-    if (data.method == 'resize') {
+    if (data.type === 'embed-size' && data.sentinel === 'amp') {
       const height = data.height;
       this.getVsync().measure(() => {
```
Comment (src-code) [inline extensions/amp-graphiq/0.1/amp-graphiq.js]: In PR #8000 @jridgewell mentioned that this use of `this.getVsync().measure()` isn't necessary, I just hadn't cleaned it up yet

**Comment by iampueroo** [comment]
@src-code, @jridgewell fixed the issues from the code comments.

Re frozen:

> So for a given widget id, could you load either the frozen or unfrozen version of it at any given time based on what domain you load it from, assuming that it's been frozen in configuration

Yes, the important point being **assuming that it's been frozen in configuration**. The only time something would "break" is if someone happens to have an unfrozen widget id and creates a `data-frozen` `<amp-graphiq>` component by themselves, since we don't freeze widgets on request and the request doesn't hit our servers (for stability - once frozen they'll never go down unless S3 does...). 

Unfortunately, as you said, we don't have a way know this on the front-end without some additional request, so I do think the `data-frozen` attribute is necessary.

**Comment by src-code** [comment]
> Unfortunately, as you said, we don't have a way know this on the front-end without some additional request, so I do think the data-frozen attribute is necessary.

Ok, makes sense to me - could you add user documentation for this attribute, and make mention of how all this works?  

Also, did you figure out the `Cannot resize element and overflow is not available` warning you saw earlier?  Wonder if @jridgewell would have any insights into that one...

**Comment by iampueroo** [comment]
Cool, will do. 

And I got the warning to go away by adding an overflow div à la [amp-iframe docs](https://github.com/ampproject/amphtml/blob/master/extensions/amp-iframe/amp-iframe.md#iframe-resizing), but didn't commit it. Wasn't sure if it was intended to be required  for a custom element.

File: extensions/amp-graphiq/amp-graphiq.md [hunk 4]
```diff
@@ -63,6 +63,10 @@ The Graphiq widget id.
 
 The `data-href` attribute can optionally be defined to override the default href used for the "See more details" link which appears in the widget. The default behavior is to link to an expanded view of the widget on Graphiq's site.
 
+**data-frozen** (optional)
+
+The existence of the `data-frozen` attribute will set the subdomain of the widget to `sw.graphiq.com`, indicating that the widget was frozen. A frozen widget's data does not update - it reflects whatever the data was at the time of freezing. `data-frozen` only changes the subdomain, **it does not freeze the visualization**. Thus, it can only be set for an already frozen widget, usually provided directly from Graphiq (which executes the freezing).
+
```
Comment (iampueroo) [inline extensions/amp-graphiq/amp-graphiq.md]: @src-code hows this?

Comment (src-code) [inline extensions/amp-graphiq/amp-graphiq.md]: 👍

**Comment by src-code** [comment]
> And I got the warning to go away by adding an overflow div à la amp-iframe docs, but didn't commit it. Wasn't sure if it was intended to be required for a custom element.

Yeah, that seems odd to me - @jridgewell, any idea why this custom extension would want an overflow element like that?

**Comment by jridgewell** [comment]
You should only get that warning when the resize is denied and there's not an overflow element. But this is a user concern, not a developer one.

**Comment by src-code** [comment]
> You should only get that warning when the resize is denied and there's not an overflow element. But this is a user concern, not a developer one.

Ok, makes sense - in this case, I think this pull is ready to merge.

**Pull request merged**